Issue with exchange type reading from properties file

Hi team,
Getting issue with the code described below:


    <!--dependency for rmq implementation-->
    <dependency>
      <groupId>org.springframework.cloud</groupId>
      <artifactId>spring-cloud-stream-binder-rabbit</artifactId>
      <version>4.0.3-SNAPSHOT</version>
      <exclusions>
        <exclusion>
          <groupId>org.springframework.cloud</groupId>
          <artifactId>spring-cloud-stream-binder-rabbit-core</artifactId>
        </exclusion>
        <exclusion>
          <groupId>org.springframework.cloud</groupId>
          <artifactId>spring-cloud-stream</artifactId>
        </exclusion>
      </exclusions>
    </dependency>

    <!-- https://mvnrepository.com/artifact/org.springframework.cloud/spring-cloud-stream-binder-rabbit-core -->
    <dependency>
      <groupId>org.springframework.cloud</groupId>
      <artifactId>spring-cloud-stream-binder-rabbit-core</artifactId>
      <version>4.0.3-SNAPSHOT</version>
    </dependency>


    <!-- https://mvnrepository.com/artifact/org.springframework.cloud/spring-cloud-stream -->
    <dependency>
      <groupId>org.springframework.cloud</groupId>
      <artifactId>spring-cloud-stream</artifactId>
      <version>4.0.3-SNAPSHOT</version>
    </dependency>


we are using exchangeType fanout but it is always taking topic. But same properties are working fine in non-native app
```yaml
spring:
  cloud:
    stream:
      rabbit:
        bindings:
          emailevent-out-0:
            producer:
              routingKeyExpression: '''events.in'''
              exchangeType: direct
              use-confirm-header: true
          audit-out-0:
            producer:
              exchangeType: fanout
              use-confirm-header: true
      default-binder: binder1
      binders:
        binder1:
          type: rabbit
          environment:
            spring:
              rabbitmq:
                virtual-host: virtual-host1
                host: localhost
                port: 5672
                username: guest
                password: guest
                ssl:
                  enabled: false
        binder2:
          type: rabbit
          environment:
            spring:
              rabbitmq:
                virtual-host: virtual-host2
                host: localhost
                port: 5672
                username: guest
                password: guest
                ssl:
                  enabled: false
      bindings:
        audit-out-0:
          producer:
            error-channel-enabled: true
          binder: binder1
          destination: domain.events
        emailevent-out-0:
          binder: binder2
          destination: domain.events2
```
error:
inequivalent arg 'type' for exchange 'domain.events' in vhost 'virtual-host1': received 'topic' but current is 'fanout'

Kindly help in resolution of the same.

@jhornakova thanks....I added those and the above error is done but now I am getting the below as my properties are:
emailevent-out-0:
            producer:
              routingKeyExpression: '''events.in'''
              exchangeType: direct
              use-confirm-header: true

@olegz @sobychacko 
Could you please help as the error post the above solution is as :
Caused by: org.springframework.boot.context.properties.bind.BindException: Failed to bind properties under 'spring.cloud.stream.rabbit.bindings.emailevent-out-0.producer.routing-key-expression' to org.springframework.expression.Expression

Let me know if anything else is needed.
Thanks

--- a/core/spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultBinderFactory.java
+++ b/core/spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultBinderFactory.java
@@ -543,9 +543,11 @@ public class DefaultBinderFactory implements BinderFactory, DisposableBean, Appl
 	 */
 	GenericApplicationContext createUnitializedContextForAOT(String configurationName,
 			Map<String, Object> binderProperties, BinderConfiguration binderConfiguration) {
-
 		GenericApplicationContext binderContext = new GenericApplicationContext();
-
+		// Set the conversion service on the binder producing context to handle complex properties
+		if (this.context != null) {
+			binderContext.getBeanFactory().setConversionService(this.context.getBeanFactory().getConversionService());
+		}
 		MapPropertySource binderPropertySource = new MapPropertySource(configurationName, binderProperties);
 		binderContext.getEnvironment().getPropertySources().addFirst(binderPropertySource);
 		binderContext.setDisplayName(configurationName + "_context");

